### PR TITLE
[Relay][ADT]Static Tensor Array

### PR DESCRIPTION
**Pull request by kevinthesun**
Add static tensor array for fixed rank tensor array. With this change, we can more easily do type inference and optimization for most tensor array use cases.

Thanks for @wweic working on the base infra of StaticTensorArrayOps class.

@wweic @zhiics @yongwww @icemelon9

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
There are lots of AST written by hand. Can you try using the relay parser?

**Comment by masahi** [comment]
Is the constraint only fixed "rank"? I have a use case for fixed "shape" tensor array (a tighter constraint)
https://discuss.tvm.ai/t/a-tensor-array-usage-question-stacking-a-list-of-fixed-sized-tensors/5565

**Comment by kevinthesun** [comment]
> There are lots of AST written by hand. Can you try using the relay parser?

This PR follows the patten of current Tensor Array. What do you mean by using relay parser?

**Comment by kevinthesun** [comment]
> https://discuss.tvm.ai/t/a-tensor-array-usage-question-stacking-a-list-of-fixed-sized-tensors/5565



> https://discuss.tvm.ai/t/a-tensor-array-usage-question-stacking-a-list-of-fixed-sized-tensors/5565

@masahi fixed shape is a subset of fixed rank and supported in this PR. For stack operator, even if we have all tensors in the array with the same static shape, the first axis output shape will still be Any().

**Comment by masahi** [comment]
@kevinthesun Great! I've been waiting for this (thanks @wweic). I think the first output axis of stack being Any makes sense and it is no problem for me, assuming it is the length of the array.

> @MarisaKirisame This PR follows the patten of current Tensor Array. What do you mean by using relay parser?

I think Marisa is talking about writing the new ADT and functions in the "Relay language" itself and use the parser to make it available to python, similar to the way List ADT and its helper functions are implemented in Prelude.

**Comment by kevinthesun** [comment]
> I think Marisa is talking about writing the new ADT and functions in the "Relay language" itself and use the parser to make it available to python, similar to the way List ADT and its helper functions are implemented in Prelude.

I see. My understanding for this is List ADT primitives are small and mature enough to be implemented in prelude. We can also move tensor array related primitives into prelude when they becomes more mature. In addition, we might want to move the generic tensor array as well, not just static tensor array. Does this make sense?

**Comment by masahi** [comment]
> Does this make sense?

Yes. Since the generic tensor array is already implemented in this way, I also think it is better to let this in first and later port all tensor array stuff to Relay lang. There could be common code between generic/static that should be refactored in the process.

**Comment by wweic** [comment]
@MarisaKirisame The reason we can not use relay parser is that we want to dynamically generate the operators for specific shape while we convert the TF model to relay IR. Current relay text parser can not easily do that.

File: python/tvm/relay/prelude.py [hunk 1]
```diff
@@ -27,6 +27,538 @@
...
+            self._get_adt_by_shape([Any(),] + list(self.shape))
+        t = Var("t")
+        case = Clause(PatternConstructor(origin_tensor_constructor, [PatternVar(t)]),
+                      tensor_constructor(op.expand_dims(t, 0, 1)))
+
+        self.prelude.mod[expand_dims_var] = \
+            Function([x], Match(x, [case], False), tensor_type_var(), [])
+
+    def define_tensor_array_read(self):
+        """Defines a function to get the head of a list. Assume the list has at least one
```
Comment (wweic) [inline python/tvm/relay/prelude.py]:
```suggestion
        """Defines a function to get the nth element of a list. Assume the list has at least one
```

File: python/tvm/relay/prelude.py [hunk 2]
```diff
@@ -27,6 +27,538 @@
...
+            Function([tensor_var], helper_var(const(0), unstack_length, tensor_var),
+                     self.prelude.l(reduced_tensor_type_var()), [])
+
+    def define_tensor_array_scatter(self, indices_shape=None, force_update=False):
+        """Defines a function to scatter the values of a tensor_t in indices of a tensor array.
+        tensor_array_scatter(ta, indices, value) :
+            list[tensor_t] -> Tensor[(Any), int32] -> tensor_t -> list[tensor_t]
+
+        Set static indices shape by specifying indices_shape.
+        Set for_update to get static indices shape operator.
```
Comment (wweic) [inline python/tvm/relay/prelude.py]:
```suggestion
        Set force_update to get static indices shape operator.
```

File: python/tvm/relay/prelude.py [hunk 3]
```diff
@@ -27,6 +27,538 @@
...
+            return
+
+        concat_name = self.get_name("tensor_concatenate")
+        concat_var = self._create_global_var(concat_name)
+        setattr(self.prelude, concat_name, concat_var)
+        output_shape = [Any(),] + list(self.shape[1:])
+        tensor_type_var, tensor_constructor = \
+            self._get_adt_by_shape(output_shape)
+
+        origin_tensor_constructor = self.get_var('tensor_constructor')
```
[the remark this answers is not in the document]
Comment (kevinthesun) [reply python/tvm/relay/prelude.py]: The reason is that input shape and output shape can be different, and we need different constructor. Similar idea in take operator.

**Comment by kevinthesun** [comment]
@wweic Comments addressed. PTAL.

File: python/tvm/relay/prelude.py [hunk 4]
```diff
@@ -217,7 +218,7 @@ def define_tensor_array_unstack(self):
         tensor_array_unstack_tensor(t) : tensor_t -> list[tensor_t]
         """
         ndim = len(self.shape)
-        # Skip scalar case
+        # We don't register unstask for scalar tensor array
```
Comment (wweic) [inline python/tvm/relay/prelude.py:221]:
```suggestion
        # We don't register unstack for scalar tensor array
```

File: python/tvm/relay/prelude.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wweic) [reply python/tvm/relay/prelude.py:143]: Thanks!

File: python/tvm/relay/prelude.py [hunk 5]
```diff
@@ -666,6 +1200,20 @@ def get_var(self, canonical, dtype):
         name = self.get_name(canonical, dtype)
         return getattr(self, name)
 
+    def get_name_static(self, canonical, dtype, shape):
```
Comment (wweic) [inline python/tvm/relay/prelude.py:1203]: Should we unify this with the one in static tensor array ops?

**Comment by kevinthesun** [comment]
@wweic Fixed.

**Comment by kevinthesun** [comment]
Thanks @wweic @masahi @MarisaKirisame

**Pull request merged**